### ISSUE TITLE
False positives when passing an argument with a large union type

### ISSUE THREAD

**Issue by neutrinoceros** [astral-sh/ty#3672]
### Summary

Here's a self-contained version of this problem
https://play.ty.dev/2c7630a1-e2bc-46bc-894e-379643f39038
(I may be possible to reduce the size of the union while preserving the sailant properties of the example)

I expect all calls to `dump` in this script to be valid, but 2/4 are flagged as
```
Argument to function `dump` is incorrect: Expected `dict[str, dict[str, list[int | float | str] | int | float | str]] | dict[str, list[int | float | str] | int | float | str] | dict[str, dict[str, list[int | float | str]]] | dict[str, list[int | float | str]]`, found `dict[str, list[str]]` (invalid-argument-type) [Ln 38, Col 6]
```
Bit of a mouthful. Let me highlight the important part:
```
Expected `(...) | dict[str, list[int | float | str] | int | float | str] | (...)`, found `dict[str, list[str]]`
```
`list[str]` is actually compatible with `list[int | float | str] | int | float | str`, so the (correct) inferred type should absolutely be recognised as compatible with the (correct) expected type. What's surprising is how ty rejects some cases (passing `data` by variablel) and accepts different-but-seemingly-equivalent cases (passing `data` as a literal)

### Version

ty 0.0.43 (29ce3146f 2026-06-03)

**Comment by carljm** [comment]
Hi, thanks for the report! I think ty's behavior is correct here, and all type checkers error on some of these examples, though the exact details depend on how each type checker infers the type of a collection literal in different contexts.

> list[str] is actually compatible with list[int | float | str] | int | float | str

It's not valid to pass a `list[str]` where a `list[int | float | str]` is expected, because list is "invariant" (that means that `list[X]` is not a subtype of `list[Y]`, even if `X` is a subtype of `Y`). This is because the contract of `list` includes mutation methods, so if you pass your `list[str]` where a `list[int | str]` is expected, the function you passed it to could mutate it by adding an `int` to it, and since it is passed by reference, the calling code would also see that mutation, and would now have an `int` in a list it thinks is a `list[str]`.

The usual way to avoid this problem is that if your function will not actually mutate the list, have it accept e.g. `Sequence[int | float | str]` instead. Since `Sequence` only includes the immutable sequence API, it doesn't allow mutation, and therefore is "covariant". Thus, switching from `list` to `Sequence` and `dict` to `Mapping` in your annotations [solves the problem](https://play.ty.dev/0103eb28-1f71-438d-8c9e-e9958c6e26ff).

> What's surprising is how ty rejects some cases (passing data by variablel) and accepts different-but-seemingly-equivalent cases (passing data as a literal)

When the list literal is right there in the call, we are free to type it as a `list[int | float | str]` (that just happens to only have strings in it currently), so we do that in order to allow the call to succeed. When the list is in a variable, we've already decided earlier (where it was created) that it is typed as `list[str]`, and we aren't free to reinterpret its type since we don't know who else has a reference to it that is typed as `list[str]`.

(We actually could, and may decide to, look at the entire scope to decide from the start to treat `data` as `list[int | float | str]`, based on seeing the later call, but we currently don't do this; that kind of action-at-a-distance in type inference is also potentially confusing.)

**Comment by neutrinoceros** [comment]
Thanks a lot for indulging me with this amount of insight, I really appreciate it !
Indeed I failed to recognise the distinction between mutable and immutable types, both in my code and in my report.
I'm addressing the root issue at https://github.com/neutrinoceros/inifix/pull/541, but beware, it looks to me that it's also making me bump into new false positives from ty. I'll think extra hard before I report these though 😄

**Comment by neutrinoceros** [comment]
for reference the follow up issue is at: https://github.com/astral-sh/ty/issues/3684
